### PR TITLE
Accept multiple signatures/events for similarity index record and classify operations.

### PR DESCRIPTION
**Pull request by tkaemming**

### CODE REVIEW

File: src/sentry/scripts/similarity/index.lua [hunk 1]
```diff
@@ -556,14 +566,7 @@ local commands = {
                         configuration,
                         time_series,
                         signature.index,
-                        table.imap(
```
Comment (tkaemming) [inline src/sentry/scripts/similarity/index.lua]: Kind of nice that this actually simplifies/unifies this implementation.

File: src/sentry/similarity/index.py [hunk 2]
```diff
@@ -27,6 +28,38 @@ def __init__(self, cluster, namespace, signature_builder, bands, interval, reten
         self.interval = interval
         self.retention = retention
 
+    def __build_signatures(self, items):
```
Comment (tkaemming) [inline src/sentry/similarity/index.py]: This is the meat and potatoes of the Python changes, it collects an iterable of feature sets into a structure of this form:

```
Map[
  bytes,  # index id
  List[   # bands
    Map[
      bytes,  # band identifier
      int,  # number of times seen
    ]
  ]
]
```

and then transforms it into the form that is used by the Lua script, which is:

```
<number of indexes to parse> [<index id> [<number of buckets in band> [<band identifier> <number of times seen> …] …] …]
```

**Comment by getsentry-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">You should update CHANGES due to the size of this PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

File: src/sentry/similarity/features.py [hunk 3]
```diff
@@ -95,69 +93,96 @@ def extract(self, event):
...
-                    event,
-                    label,
-                    error,
-                )
-            else:
-                if features:
-                    items.append((self.aliases[label], features, ))
+        for event in events:
+            for label, features in self.extract(event).items():
+                if scope is None:
```
Comment (tkaemming) [inline src/sentry/similarity/features.py]: These assertions aren't the most elegant thing, but they work for now. Eventually it'd be nice to just collect these by `(scope, key)` and then route to the appropriate node(s) if there are more than one. This isn't a priority right now though

File: src/sentry/similarity/features.py [hunk 4]
```diff
@@ -95,69 +93,96 @@ def extract(self, event):
...
+                else:
+                    if features:
+                        items.append((self.aliases[label], features, ))
+
         return self.index.record(
-            self.__get_scope(event.project),
-            self.__get_key(event.group),
+            scope,
+            key,
             items,
```
Comment (tkaemming) [inline src/sentry/similarity/features.py]: In the future this would be probably a good idea to make a generator instead too

**Review by JTCunning — APPROVED** [review]
damn son where'd you find this

**Pull request merged**